starboard/shared/libiamf/iamf_decoder_utils.cc

```diff
@@ -0,0 +1,609 @@
+        info->mix_presentation_id = mix_presentation_id;
+        binaural_mix_selection = kBinauralMixSelectionLoudnessLayout;
+      }
+
+      // The following fields are for the LoudnessInfo
+      uint8_t info_type;
+      RCHECK(reader->Read1(&info_type));
+      // integrated_loudness and digital_loudness
+      RCHECK(reader->SkipBytes(4));
+      if (info_type & 1) {
+        // true_peak
+        RCHECK(reader->SkipBytes(2));
+      }
+      if (info_type & 2) {
+        uint8_t num_anchored_loudness;
+        RCHECK(reader->Read1(&num_anchored_loudness));
+        for (uint8_t k = 0; k < num_anchored_loudness; ++k) {
+          // anchor_element and anchored_loudness
+          RCHECK(reader->SkipBytes(3));
+        }
+      }
+      if ((info_type & 0b11111100) > 0) {
+        uint32_t info_type_size;
+        RCHECK(reader->ReadLeb128(&info_type_size));
+        // info_type_bytes
+        RCHECK(reader->SkipBytes(info_type_size));
+      }
+    }
+  }
+
+  // If the mix presentation id is unassigned at this point, the stream is
+  // stereo, or a proper mix presentation for binaural or surround preferred
+  // streams hasn't yet been parsed. Default to the first read mix
+  // presentation in case a preferred mix does not exist.
+  if (!info->mix_presentation_id.has_value()) {
+    info->mix_presentation_id = mix_presentation_id;
+  }
+
+  return true;
+}
+
+bool ParseDescriptorOBU(BufferReader* reader,
+                        IamfBufferInfo* info,
+                        const bool prefer_binaural_audio,
+                        const bool prefer_surround_audio) {
+  SB_DCHECK(reader);
+  uint8_t obu_type = 0;
+  uint32_t obu_size = 0;
+  if (!ParseOBUHeader(reader, &obu_type, &obu_size)) {
+    SB_LOG(ERROR) << "Error parsing OBU header";
+    return false;
+  }
+
+  int next_obu_pos = reader->pos() + obu_size;
+
+  std::unordered_set<uint32_t> binaural_audio_element_ids;
+  std::unordered_set<uint32_t> surround_audio_element_ids;
+  switch (static_cast<int>(obu_type)) {
+    case kObuTypeCodecConfig:
+      RCHECK(ParseCodecConfigOBU(reader, info));
+      break;
+    case kObuTypeAudioElement:
+      RCHECK(ParseAudioElementOBU(reader, info, &binaural_audio_element_ids,
+                                  &surround_audio_element_ids,
+                                  prefer_binaural_audio,
+                                  prefer_surround_audio));
+      break;
+    case kObuTypeSequenceHeader:
+      break;
+    case kObuTypeMixPresentation:
+      RCHECK(ParseMixPresentationOBU(reader, info, &binaural_audio_element_ids,
+                                     &surround_audio_element_ids,
+                                     prefer_binaural_audio,
+                                     prefer_surround_audio));
+      break;
+    default:
+      // This executes if a non descriptor OBU is read. The buffer info must be
+      // valid by this point.
+      SB_DCHECK(BufferInfoIsValid(info));
+      return true;
```

Comment:
`return BufferInfoIsValid(info);`.  So we don't return true in production for corrupted input.